clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -355,9 +365,17 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                             !type_is_prefetch(memref.data.type),
                         "Function marker misplaced between instr and memref");
     }
+    if (memref.marker.type == TRACE_TYPE_MARKER &&
+        memref.marker.marker_type == TRACE_MARKER_TYPE_FUNC_ID) {
+        shard->prev_func_id_ = memref.marker.marker_value;
+    }
     if (memref.marker.type == TRACE_TYPE_MARKER &&
         marker_type_is_function_marker(memref.marker.marker_type)) {
-        report_if_false(shard, type_is_instr_branch(shard->prev_instr_.instr.type),
+        report_if_false(shard,
+                        shard->prev_func_id_ >= TRACE_FUNC_ID_SYSCALL_BASE ||
+                            (memref.marker.marker_type == TRACE_MARKER_TYPE_FUNC_ID &&
```

Comment:
Why do we need this second part of the condition? Isn't the check on `prev_func_id_ >= TRACE_FUNC_ID_SYSCALL_BASE` enough?

If the second part is true, prev_func_id_ would be set to a value >= TRACE_FUNC_ID_SYSCALL_BASE by the if-block on L368, causing the first part to be true.

Reply:
Or: Maybe it is better to remove prev_func_id_ or make it a local var instead of a shard_data var.

Reply:
This was added in f3c36937c7c4f886540db068e5f752bc2de05cdf but it also removed the else on this if and I think that by itself may be enough.  I removed the 2nd condition.

We cannot remove prev_func_id_ nor make it a local var because there are *multiple* markers for the retaddr and args after the ID.